isobus/src/can_transport_protocol_base.cpp

```diff
@@ -0,0 +1,103 @@
+//================================================================================================
+/// @file can_transport_protocol.cpp
+///
+/// @brief Abstract base class for CAN transport protocols.
+/// @author Daan Steenbergen
+///
+/// @copyright 2023 The Open-Agriculture Developers
+//================================================================================================
+
+#include "isobus/isobus/can_transport_protocol_base.hpp"
+#include "isobus/isobus/can_internal_control_function.hpp"
+#include "isobus/utility/system_timing.hpp"
+
+namespace isobus
+{
+	TransportProtocolSessionBase::TransportProtocolSessionBase(TransportProtocolSessionBase::Direction direction,
+	                                                           std::unique_ptr<CANMessageData> data,
+	                                                           std::uint32_t parameterGroupNumber,
+	                                                           std::uint32_t totalMessageSize,
+	                                                           std::shared_ptr<ControlFunction> source,
+	                                                           std::shared_ptr<ControlFunction> destination,
+	                                                           TransmitCompleteCallback sessionCompleteCallback,
+	                                                           void *parentPointer) :
+	  direction(direction),
+	  parameterGroupNumber(parameterGroupNumber),
+	  data(std::move(data)),
+	  source(source),
+	  destination(destination),
+	  totalMessageSize(totalMessageSize),
+	  sessionCompleteCallback(sessionCompleteCallback),
+	  parent(parentPointer)
+	{
+	}
+
+	bool TransportProtocolSessionBase::operator==(const TransportProtocolSessionBase &obj) const
+	{
+		return ((source == obj.source) && (destination == obj.destination) && (parameterGroupNumber == obj.parameterGroupNumber));
+	}
+
+	bool TransportProtocolSessionBase::matches(std::shared_ptr<ControlFunction> other_source, std::shared_ptr<ControlFunction> other_destination) const
+	{
+		return ((source == other_source) && (destination == other_destination));
+	}
+
+	TransportProtocolSessionBase::Direction TransportProtocolSessionBase::get_direction() const
+	{
+		return direction;
+	}
+
+	CANMessageData &TransportProtocolSessionBase::get_data() const
+	{
+		return *data;
+	}
+
+	std::uint32_t TransportProtocolSessionBase::get_message_length() const
+	{
+		return totalMessageSize;
+	}
+
+	std::shared_ptr<ControlFunction> TransportProtocolSessionBase::get_source() const
+	{
+		return source;
+	}
+
+	float TransportProtocolSessionBase::get_percentage_bytes_transferred() const
+	{
+		return static_cast<float>(get_total_bytes_transferred()) / static_cast<float>(get_message_length());
```

Comment:
Nit: Technically to be a percentage this would have to be `* 100` otherwise it's just the fraction

Reply:
Also.... paranoid brain, should check `0 != get_message_length()` to avoid `NaN`?